presto-orc/src/main/java/com/facebook/presto/orc/writer/MapFlatColumnWriter.java

```diff
@@ -15,148 +15,568 @@
+                .putAll(valueStats)
+                .build();
+    }
+
+    @Override
+    public Map<Integer, ColumnStatistics> getColumnStripeStatistics()
+    {
+        checkState(closed);
+
+        Map<Integer, ColumnStatistics> valueStats;
+        if (valueWriters.isEmpty()) {
+            valueStats = getEmptyValueColumnStatistics();
+        }
+        else {
+            valueStats = getValueColumnStatistics(ColumnWriter::getColumnStripeStatistics);
+        }
+        return ImmutableMap.<Integer, ColumnStatistics>builder()
+                .put(nodeIndex, mergeColumnStatistics(rowGroupColumnStatistics))
+                .put(keyNodeIndex, keyManager.getStripeColumnStatistics())
+                .putAll(valueStats)
+                .build();
+    }
+
+    // Create and cache value columnStripeStatistics for cases when the there are no value writers,
+    // but we still need to write valid column stripe statistics for all nested writers.
+    private Map<Integer, ColumnStatistics> getEmptyValueColumnStatistics()
+    {
+        if (emptyValueColumnStripeStatistics == null) {
+            ColumnWriter valueWriter = valueWriterFactory.apply(0);
+            valueWriter.close();
+            // stats are the same for row group and stripe
+            emptyValueColumnStripeStatistics = valueWriter.getColumnStripeStatistics();
+        }
+        return emptyValueColumnStripeStatistics;
     }
 
-    private long writeColumnarMap(ColumnarMap columnarMap)
+    private Map<Integer, ColumnStatistics> getValueColumnStatistics(Function<ColumnWriter, Map<Integer, ColumnStatistics>> getStats)
     {
-        // TODO: implement me
-        return 0;
+        ImmutableListMultimap.Builder<Integer, ColumnStatistics> allValueStats = ImmutableListMultimap.builder();
+        for (MapFlatValueWriter valueWriter : valueWriters) {
+            Map<Integer, ColumnStatistics> valueColumnStatistic = getStats.apply(valueWriter.getValueWriter());
+            allValueStats.putAll(valueColumnStatistic.entrySet());
+        }
+
+        ImmutableMap.Builder<Integer, ColumnStatistics> columnStatistics = ImmutableMap.builder();
+        allValueStats.build().asMap().forEach((nodeIndex, nodeStats) -> {
+            ColumnStatistics mergedNodeStats = mergeColumnStatistics((List<ColumnStatistics>) nodeStats);
+            columnStatistics.put(nodeIndex, mergedNodeStats);
+        });
+
+        return columnStatistics.build();
     }
 
     @Override
-    public Map<Integer, ColumnStatistics> finishRowGroup()
+    public long writeBlock(Block block)
     {
         checkState(!closed);
-        // TODO: implement me
-        return ImmutableMap.of();
+        checkArgument(block.getPositionCount() > 0, "block is empty");
+
+        ColumnarMap columnarMap = toColumnarMap(block);
+        Block keysBlock = columnarMap.getKeysBlock();
+        Block valuesBlock = columnarMap.getValuesBlock();
+
+        long childRawSize = 0;
+        int entryPosition = 0;
```

Comment:
This could be broken, if the block passed in is view or DictionaryBlock or other changes. Inside of the loop get the offset for the entry and use it.

Reply:
Done